### ISSUE TITLE
How to reproduce the feature map visualization in the paper

### ISSUE THREAD

**Issue by TianYinchen56** [erfect2020/ContextualDeblur#4]
Hello,
<img width="1353" height="400" alt="Image" src="https://github.com/user-attachments/assets/4d8c3a1b-9d04-48bf-858c-0b334229281a" />

  I'm currently reading your paper titled ViTDeblur. I am very interested in the feature map visualizations shown in Figure 8 and the right-side images, which seem to represent the model's attention or feature activation.
  How do you convert the extracted feature maps into a heatmap and overlay them on the original image? Are there any code examples available for this?
 I have been trying to implement this , but the results are not ideal.

Thank you for your time and help!

**Comment by erfect2020** [comment]
With the help of ChatGPT, it is easy to implement.
The key idea of the visualization is cosine similarity.
If this repo help you, please cite the work.

A very simple case.

```
def chunk_cosine_sim(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
    """
    Computes cosine similarity between all possible pairs in two sets of vectors.
    Operates on chunks so no large amount of GPU RAM is required.
    :param x: an tensor of descriptors of shape Bx1x(t_x)xd' where d' is the dimensionality of the descriptors and t_x
    is the number of tokens in x.
    :param y: a tensor of descriptors of shape Bx1x(t_y)xd' where d' is the dimensionality of the descriptors and t_y
    is the number of tokens in y.
    :return: cosine similarity between all descriptors in x and all descriptors in y. Has shape of Bx1x(t_x)x(t_y)
    """
    result_list = []
    num_token_x = x.shape[2]
    for token_idx in range(num_token_x):
        token = x[:, :, token_idx, :].unsqueeze(dim=2)  # Bx1x1xd'
        result_list.append(torch.nn.CosineSimilarity(dim=3)(token, y))  # Bx1xt
    return torch.stack(result_list, dim=2)  # Bx1x(t_x)x(t_y)
```

**Comment by TianYinchen56** [comment]
Thank you for your wonderful paper and your previous helpful response,and I have two questions that I would be grateful for your insights on.

Visualization Method: In the visualization of the feature maps and attention maps (e.g., in Figure 8), the feature maps are displayed at the original image resolution. My understanding is that the Vision Transformer processes features at a lower resolution (e.g., 14×14 for a 224×224 image). Is it correct that these visualizations are generated by upsampling the low-resolution feature maps back to the original image size? This seems like an informative but approximate representation, and I wanted to confirm if this is the case.

Inspiration for BST Classification Loss: My background is in NeRF and 3D Gaussian Splatting, and I am keen to understand the motivations behind your design choices. I found the Blur-Sharp Translator (BST) classification loss to be a particularly interesting and effective component of your method. Could you please share what inspired you to introduce this auxiliary classification task? What problem were you aiming to solve, and how did you come to the conclusion that a classification loss would be the right approach to guide the model's feature learning?

Thank you for your time and for sharing your valuable work with the community.

Best regards

**Comment by erfect2020** [comment]
1. There are two solutions for that: a) Upsample the feature map. b) Upsample the original input image.

2. Regarding the BST design principle — it was inspired by the idea of hierarchical world models, meaning multilevel representations that support abstraction across different time scales. This concept is discussed in Yann LeCun’s position paper A Path Towards Autonomous Machine Intelligence (2022), where he proposes “hierarchical joint embedding architectures” and multiscale predictive world models as core components for autonomous agents. At the time this work was done, hierarchical representation learning was getting a lot of attention in the community, and it’s still a central theme in LeCun’s vision.

**Comment by TianYinchen56** [comment]
I'm certain that your work is a true feat of engineering artistry. Having read your paper several times, I can really appreciate the elegance of the design and the thinking behind it.
